fix #30480 torch.normal shape checking is broken

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32243 fix #30480 torch.normal shape checking is broken**

Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

Commit:
fix #30480 torch.normal shape checking is broken

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 289b5ad4:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=289b5ad40f7e228ff8b8e09d7897154deb6632e9).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4260033) pytorch_linux_xenial_py2_7_9_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86657293) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86657293))
<details>
<summary>
<code>Jan 15 21:59:31 RuntimeError: test_dataloader failed!</code>
</summary>

```
Jan 15 21:59:31 Ran 62 tests in 88.711s 
Jan 15 21:59:31  
Jan 15 21:59:31 FAILED (failures=1, skipped=8) 
Jan 15 21:59:31  
Jan 15 21:59:31 Generating XML reports... 
Jan 15 21:59:31 Traceback (most recent call last): 
Jan 15 21:59:31   File "test/run_test.py", line 456, in <module> 
Jan 15 21:59:31     main() 
Jan 15 21:59:31   File "test/run_test.py", line 449, in main 
Jan 15 21:59:31     raise RuntimeError(message) 
Jan 15 21:59:31 RuntimeError: test_dataloader failed! 
Jan 15 21:59:31 + cleanup 
Jan 15 21:59:31 + retcode=1 
Jan 15 21:59:31 + set +x 
Jan 15 21:59:31 =================== sccache compilation log =================== 
Jan 15 21:59:31 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Jan 15 21:59:31  
Jan 15 21:59:31 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 15 21:59:31 Compile requests                 46 
Jan 15 21:59:31 Compile requests executed        27 
Jan 15 21:59:31 Cache hits                       20 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zou3519 reviewed (changes requested):
Please add tests that check the behavior of torch.normal on the three behaviors you listed for CPU and CUDA tensors

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Test case is added for only gpu case. gpu normal uses new tensor iterator while cpu version is still using legacy code(TH). We need to make sure gpu version behave correctly and migrate the cpu version later to use new tensor iterator.

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Test case is added for only gpu case. gpu normal uses new tensor iterator while cpu version is still using legacy code(TH). We need to make sure gpu version behave correctly and migrate the cpu version later to use new tensor iterator.

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Test case is added for only gpu case. gpu normal uses new tensor iterator while cpu version is still using legacy code(TH). We need to make sure gpu version behave correctly and migrate the cpu version later to use new tensor iterator.

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -730,7 +730,14 @@ Tensor& normal_out_cuda(Tensor& output, const Tensor& mean, const Tensor& std, G
-  output.mul_(std).add_(mean);
+  if (mean.numel() == std.numel() && !at::is_expandable_to(std.sizes(), mean.sizes())) {
+    TORCH_WARN_ONCE("Reshape std to the same shape of mean due to not broadcastable");
+    //Reshape std to the same shape mean if they have same number of element but not broadcastable
```
you don't need a comment if the warning says the same thing as the comment.

glaringlee replied:
Removed

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -730,7 +730,14 @@ Tensor& normal_out_cuda(Tensor& output, const Tensor& mean, const Tensor& std, G
   // Consequently, the computation performed is mean + mean * std instead of mean + output * std
-  output.mul_(std).add_(mean);
+  if (mean.numel() == std.numel() && !at::is_expandable_to(std.sizes(), mean.sizes())) {
+    TORCH_WARN_ONCE("Reshape std to the same shape of mean due to not broadcastable");
```
this isn't really a good description of the situation.  It's not clear from the user's perspective how this information is actionable.

How about something like:
"std and mean have the same number of elements, but are not broadcastable.  This was previously a supported mode of operation, but is now deprecated and the support will be removed in a later release.  Note that the current implementation reshapes std to the shape of mean, which may be incur data copies.  Please ensure that std and mean are broadcastable to avoid these issues."

gchanan replied:
I also don't think `is_expandable_to` is the correct function to call -- that  name suggests directionality, but you really just want to know if the two tensors are broadcastable.

gchanan replied:
please add a test of this case so we can be sure we don't make this mistake.

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,34 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
 Tensor& normal_out_cuda(Tensor& output, const Tensor& mean, const Tensor& std, Generator* gen) {
+  bool expandable = are_expandable(mean.sizes(), std.sizes());
+  // allow resize if output is an empty tensor
+  if (output.sizes().equals({0})) {
```
I forgot to think about backwards compatibility when suggesting this change. It looks like torch.normal (on CPU) allows resize of the output no matter what its size is. The old (THC) torch.normal on CUDA probably also allows the same thing.

However, the current torch.normal (on CUDA) doesn't allow for any resizing at all, and we've had this behavior for 8 months now (and no one has noticed yet...). Given that, I think it's fine if we don't enable the legacy resize behavior of "resize everything".

gchanan replied:
I don't know what "resize everything" means.  But I don't understand how the "same element, non-broadcastable" form works because add_ and mul_ don't support that.

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,58 @@ def ref_log_prob(idx, x, log_prob):
+        self.assertEqual(torch.normal(tensor23, tensor6).size(), (2, 3))
+        self.assertEqual(torch.normal(tensor6, tensor23).size(), (6,))
+        # std and mean have different number of element
+        with self.assertRaises(RuntimeError):
```
We should use `self.assertRaisesRegex(RuntimeError, "snippet of error message")` instead to ensure that the error message is readable

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,58 @@ def ref_log_prob(idx, x, log_prob):
+        tensor21 = torch.rand(2, 1)
+        tensor6 = torch.rand(6)
+        tensor4 = torch.rand(4)
+        # normal case
```
Probably better to name this something else, like "inputs have same size" to avoid confusion with the  op name

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,58 @@ def ref_log_prob(idx, x, log_prob):
+
+    @unittest.skipIf(not TEST_CUDA, "CUDA not found")
+    def test_normal_shape_gpu(self):
+        tensor12 = torch.rand(1, 2, device='cuda')
```
(This isn't blocking) Is it possible to share code between `test_normal_shape_cpu` and `test_normal_shape_gpu`? 

On CPU, we support: (1) std/mean have same sizes, (2) std/mean have different sizes
On CUDA, we support: (1) std/mean have same sizes, (2) std/mean have different sizes, (3) std/mean are broadcastable

We could probably toss (1) and (2) into a helper function, but I'm not sure how much cleaner it would make this implementation. The reason why the tests are like this is because we implement CPU and CUDA separately...

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,58 @@ def ref_log_prob(idx, x, log_prob):
+            torch.normal(tensor23, tensor4)
+
+    @unittest.skipIf(not TEST_CUDA, "CUDA not found")
+    def test_normal_shape_gpu(self):
```
Should call this `test_normal_shape_cuda` instead, `gpu` can refer to things that aren't nvidia gpus

zou3519 reviewed (commented):
approach looks fine, I had some suggestions for the testing

gchanan commented on aten/src/ATen/ExpandUtils.h:
```diff
@@ -174,4 +174,19 @@ static inline bool is_expandable_to(IntArrayRef shape, IntArrayRef desired) {
 }
 
+// True if the input two shapes are broadcastable from one to the other in any direction
+static inline bool are_expandable(IntArrayRef shape1, IntArrayRef shape2) {
```
infer_size already does this; if the error message is a problem, refactor that code instead of introducing code that does the same logic.

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,34 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
+  if (output.sizes().equals({0})) {
+    if (expandable) {
+      auto shape = at::infer_size(mean.sizes(), std.sizes());
+      at::native::resize_(output, shape, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
you don't need to use the LEGACY_CONTIGUOUS_MEMORY_FORMAT assuming normal_cuda_, mul_, add_ support any memory format (you can just check or ask @VitalyFedyunin).

glaringlee replied:
Will check with Vitaly. Using LEGACY_CONTIGUOUS_MEMORY_FORMAT here is because the output tensor is empty at this point, we actually don't know what kind of memory format it should be allocated. So probably LEGACY_CONTIGUOUS_MEMORY_FORMAT is the best choice. And all the non-out version normal_cuda function all allocate output tensor using LEGACY_CONTIGUOUS_MEMORY_FORMAT.  I think this is consistent and backward compatible.

gchanan replied:
but LEGACY_CONTIGUOUS_MEMORY_FORMAT is exactly equivalent to CONTIGUOUS_FORMAT, it just means we haven't actually checked that it's safe to make it not contiguous.

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,34 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
+      at::native::resize_(output, shape, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+    }
+    else {
+      at::native::resize_(output, mean.sizes(), LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
You don't want to resize the output if the op isn't going to succeed because resize is a destructive operation.  It is true that the previous would potentially write data and then fail, but it wouldn't resize.

glaringlee replied:
Will think about this more. Probably create a temp tensor doing the op in this case. And if success, then doing a inplace copy to the output tensor.

gchanan replied:
I don't think this is correct.  We define out variants of this function, so you want to get the resize behavior correct.

I don't follow the argument about the non-out variant either.  What's the point of passing it to normal_cuda anyway -- you know it will fail later.

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,41 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
 
 Tensor& normal_out_cuda(Tensor& output, const Tensor& mean, const Tensor& std, Generator* gen) {
+  bool expandable = are_expandable(mean.sizes(), std.sizes());
+  bool empty_output = output.sizes().equals({0});
```
I think you should just check if the number of elements in output is 0.

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,41 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
+
+  if (expandable) {
+    auto shape = at::infer_size(mean.sizes(), std.sizes());
+    TORCH_CHECK(empty_output || output.sizes().equals(shape), "output size is not the same as broadcast size of mean and std");
```
nit: say "broadcasted size of mean and std" instead of "broadcast size"

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,41 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
+    }
+  }
+  else {
+    TORCH_CHECK(mean.numel() == std.numel(), "mean and std should have same number of element when nonexpandable");
```
some more detail here would be good.

For example:
- explain the deprecated support for non-broadcastable shapes
- do we print why the tensors weren't broadcastable?  (we usually do for cases where we are actually checking that).
- do we print the number of elements for both tensors?

gchanan reviewed (approved):
I didn't look at the tests.  The error messages will be more understandable if you move up the deprecated warning.

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -724,13 +724,41 @@ Tensor& normal_out_cuda(Tensor& output, double mean, const Tensor& std, Generato
+
+  if (expandable) {
+    auto shape = at::infer_size(mean.sizes(), std.sizes());
+    TORCH_CHECK(empty_output || output.sizes().equals(shape), "output size is not the same as broadcasted size of mean and std");
```
nit (not blocking): It's nice to report the sizes (output size, size of mean, size of std) so users can correlate the function call back to the inputs.

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,39 @@ def ref_log_prob(idx, x, log_prob):
 
         self._check_log_prob(Normal(loc, scale), ref_log_prob)
 
+    def test_normal_shape(self):
```
nit (not blocking): I prefer the testing of this function to be in test_torch.py, because torch.normal is a function on torch.*, not a function on torch.distributions. But that's a pretty minor detail; don't worry about it if you don't want to

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,39 @@ def ref_log_prob(idx, x, log_prob):
+            self.assertEqual(torch.normal(tensor120, tensor2345).size(), (120,))
+            self.assertEqual(torch.normal(tensor2345, tensor120).size(), (2, 3, 4, 5))
+            # inputs are non-expandable tensors and they don't have same number of elements
+            with self.assertRaises(RuntimeError):
```
Is the error message different on CPU/CUDA? It would be nicer if we could test what the error message is here.

glaringlee replied:
They are very similar, I changed the code to test the error message.

zou3519 commented on test/test_distributions.py:
```diff
@@ -1603,6 +1603,39 @@ def ref_log_prob(idx, x, log_prob):
+            # inputs are non-expandable tensors and they don't have same number of elements
+            with self.assertRaises(RuntimeError):
+                torch.normal(tensor2345, tensor4)
+
```
We should add tests for the out= variant. Namely, that if the output tensor's shape does not match the desired shape, then torch.normal should raise an error.

I expect this logic to only exist on CUDA (since the CPU version is using TH and TH didn't care about those semantics...)

glaringlee replied:
Added

zou3519 reviewed (commented):
We should add a test for the resizing behavior (that torch.normal(..., out=output) on CUDA yells if the output doesn't have the same size as the calculated size). Other than that, the tests lgtm

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -5960,6 +5960,52 @@ def test_tensor_grad_warnings(self):
+            self.assertEqual(torch.normal(tensor2345_non_contiguous, tensor2345).size(), (2, 3, 4, 5))
+            self.assertEqual(torch.normal(tensor2345, tensor2345_channels_last).size(), (2, 3, 4, 5))
+            self.assertEqual(torch.normal(tensor2345_non_contiguous, tensor2345_channels_last).size(), (2, 3, 4, 5))
+            # scalar case
```
nit: Some newlines here between groups of tests would make the tests a bit more readable:

```
# inputs have same size
...

# scalar case
...
```

Alternatively you can split these things up into their own test cases

glaringlee replied:
done

zou3519 commented on test/test_torch.py:
```diff
@@ -5960,6 +5960,52 @@ def test_tensor_grad_warnings(self):
+            # inputs are non-expandable tensors and they don't have same number of elements
+            with self.assertRaisesRegex(RuntimeError, "inconsistent tensor"):
+                torch.normal(tensor2345, tensor4)
+            # output and inputs are size sompatialbe
```
sompatiable -> compatible

zou3519 commented on test/test_torch.py:
```diff
@@ -5960,6 +5960,52 @@ def test_tensor_grad_warnings(self):
+            # output and inputs are size sompatialbe
+            self.assertEqual(torch.normal(tensor2345, tensor2345, out=output2345).size(), (2, 3, 4, 5))
+            # output and inputs are not size compatible
+            # CUDA only since CPU version is written in legacy TH, it doesn't care about the size compatiable
```
compatiable -> compatible

zou3519 commented on test/test_torch.py:
```diff
@@ -5960,6 +5960,52 @@ def test_tensor_grad_warnings(self):
+            # output and inputs are not size compatible
+            # CUDA only since CPU version is written in legacy TH, it doesn't care about the size compatiable
+            if device == 'cuda':
+                with self.assertRaisesRegex(RuntimeError, "inconsistent tensor"):
```
This checks that the following code raises a RuntimeError. It doesn't check that every line raises an error.

To do that, we'd do something like:
```
with self.assertRaisesRegex(RuntimeError, "inconsistent tensor"):
    torch.normal(tensor2345, tensor2145, out=output345)
with self.assertRaisesRegex(RuntimeError, "inconsistent tensor"):
    torch.normal(tensor2345, tensor120, out=output345)
```

zou3519 commented on test/test_torch.py:
```diff
@@ -5960,6 +5960,52 @@ def test_tensor_grad_warnings(self):
+                    torch.normal(tensor2345, tensor2145, out=output345)
+                    # inputs are not expandable but reshapeable, output size is not the same as mean
+                    torch.normal(tensor2345, tensor120, out=output345)
+
```
Probably good to add another test case where the the `out=` argument is compatible so we're sure that that still works.

glaringlee replied:
The compatible case is already added above, for both CPU and CUDA. see lines above this segment.

zou3519 reviewed (approved):
some minor comments on the tests, after that it looks good to go

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

zou3519 reviewed (approved):
LGTM

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

Commit:
Update on "fix #30480 torch.normal shape checking is broken"


Following what @gchanan proposed in #30480 
- If the (logical) shapes of mean and std are broadcastable, we broadcast them for the output
  Done in tensor iterator already.
- If the (logical) shapes of mean and std are not broadcastable and they have the same number of elements, we fall back to the old behavior (pick the shape of mean)
  Done by reshape std to the same shape of mean.
- If the (logical) shapes of mean and std are not broadcastable and don't have the same number of elements, we error out.
  Done by tensor iterator already.

Differential Revision: [D19417087](https://our.internmc.facebook.com/intern/diff/D19417087)

Currently, cpu version is still written in legacy code (TH), it doesn't support broadcasting. This will be solved once it is migrated using new tensor iterator like gpu version. Test cases for cpu and gpu are separated provided at this time. 
I am making sure gpu version behave correctly at this point, and cpu version behave all the same as gpu except tensor broadcasting support.

[ghstack-poisoned]

zou3519 commented:
@glaringlee I am reverting this because it broke some tests on master: https://circleci.com/gh/pytorch/pytorch/4430687?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console (see https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master). It looks like there may have been a crossed merge, please feel free to resubmit this PR
